File: src/endpoints/app/get_item.py
from typing import Optional
from src import output
from src.libs.db import DB, Table
from src.modules.verify import checking_token
from src.modules.mod import MOD
from . import __types__ as types

async def get_item(params: dict = {}, req = None, _db: DB = None, _check_token = True):

    # set values
    result = None
    db = _db if _db else DB().connect()

    try:
        # set params
        params = types.GetItem(**params)

        # checking token
        if _check_token: checking_token(req, db)

        # set srl
        srl: Optional[int] = None
        code: Optional[str] = None
        try: srl = int(params.srl)
        except ValueError: code = str(params.srl)

        # set fields
        fields = params.fields.split(',') if params.fields else None

        # set mod
        mod = MOD(params.mod or '')

        # set where
        where = []
        if srl: where.append(f'and srl={srl}')
        if code: where.append(f'and code LIKE \'{code}\'')

        # get item
        data = db.get_item(
            table_name=Table.APP.value,
            where=where,
            fields=fields,
        )
        if not data: raise Exception('Item not found', 204)

        # MOD / count-nest
        if mod.check('count-nest'):
            from ..nest import __libs__ as nest_libs
            data['count_nest'] = nest_libs.get_count(db, [ f'app_srl = {data['srl']}' ])

        # MOD / count-article
        if mod.check('count-article'):
            from ..article import __libs__ as article_libs
            data['count_article'] = article_libs.get_count(db, [ f'app_srl = {data['srl']}' ])

        # set result
        result = output.success({
            'message': 'Complete get app item.',
            'data': data,
        }, _req=req)
    except Exception as e:
        result = output.exc(e, _req=req)
    finally:
        if not _db and db: db.disconnect()
        return result

File: src/routes/app/get-item.ts
import DB, { db } from '@/classes/DB'
import ServiceError from '@/classes/ServiceError'
import MOD from '@/classes/MOD'
import * as nestHelper from '@/routes/nest/__helper'
import * as helper from './__helper'
import type { AppModel } from './__model'

type GetItemParams = {
  srl?: number
  code?: string
  query: AppModel['getItemQuery']
}

export default async function getItem({ srl, code, query }: GetItemParams)
{
  try
  {
    // set field
    const _field = query.field ? query.field.split(',') : ''

    // set where
    let _where = ''
    if (srl) _where = `srl = ${srl}`
    else if (code) _where = `code LIKE \'${code}\'`

    // get item
    const item = db.getData({
      table: DB.TABLE.APP,
      field: _field,
      where: _where,
    })
    if (!item.data)
    {
      throw new ServiceError('App data not found.', { status: 204 })
    }

    // set data
    let data: ZZ = {
      ...(item ? item.data : {}) as ZZ,
    }
    // set MOD
    const _mod: MOD = new MOD(query.mod)
    // MOD / count-nest
    if (_mod.check('count-nest'))
    {
      data.count_nest = nestHelper.count({
        where: `app_srl = ${item.data.srl}`,
      })
    }
    // MOD / count-article
    if (_mod.check('count-article'))
    {
      data.count_article = helper.countArticle(item.data.srl)
    }

    return data
  }
  catch (_e: any)
  {
    throw new ServiceError('Failed to get App.', {
      status: _e.status,
      text: _e.message,
      cause: _e,
    })
  }
}
